ir/indexed_vector.h

```diff
@@ -212,6 +212,9 @@ class IndexedVector : public Vector<T> {
                       "invalid element %1%", el);
         }
     }
+
+    DECLARE_TYPEINFO_WITH_NESTED_TYPEID(IndexedVector<T>, NK_IndexedVectorT, T::static_typeId(),
```

Comment:
Is it necessary to use the nested type ID? The name used for the ID derivation does contain the template argument anyway, right?

Reply:
I want to develop the foundation of possible improved usage in the future. In particular, static polymorphism. With this PR it is possible to do things like this:

```c++
switch (node->typeId()) {
  case NK_Type:
    ...
  case NK_SomethingElse:
    ...
  default:
    ...
}
```
without relying on double-dispatch in Visitors and lots of overheads involved. But what can we do with `Vector<T>` and `IndexedVector<T>`? Can we do something similar here? This is nested typeid is for. Essentially it is split into two pieces: 8 bit discriminator and 56-bits of "real" typeid. `0x00` discriminator is currently reserved for `IR::Node` hierarchy, `0xFF` is for hash-produced values. Correspondingly `0x01` is for `Vector` and `0x02` is for `IndexedVector`.

Now it is possible to:
  - Determine if we are having a particular `Vector<SomeClass>` via checking the full typeid
  - Determine if this is a generic `Vector<T>` via checking discriminator
  - Perform the switch as above via `T`'s typeid from `Vector<T>` or `IndexedVector<T>`.

Hope it makes the usage case of nested typeid's clear

Reply:
And note that the implementation supports two ways to specify typeid:
  - Explicit. And this is used for IR::Node and friends. We rely on `ir-generator` to generate `NodeKind` enum for us.
  - Implicit. Autogenerated from the hash of type name

Reply:
Yes, though we know which classes are abstract and which are concrete via that tree macro. Still, it's something that one could consider.